### PR TITLE
Added sorting of warnings during import and displaying them with counts

### PR DESCRIPTION
**Pull request by j-hudecek**
This pull request changes the way warnings are handled during data import. ProgressMonitor now stores them in a TreeSet (stores each warning only once in order in log(n)). At the end of import they are output sorted by alphabet with a count of how many times they occured during import. Example:

    Gene not found:  TXNP7 [0]. Ignoring it and all mutation data associated with it!; 2x
    Gene not found:  U3 [0]. Ignoring it and all mutation data associated with it!; 12x
    Gene not found:  U40455.1 [0]. Ignoring it and all mutation data associated with it!; 2x
    Gene not found:  U66059.29 [0]. Ignoring it and all mutation data associated with it!; 2x
    Gene not found:  U8 [0]. Ignoring it and all mutation data associated with it!; 4x

This limits the output of the command and prints similar pseudogenes close to each other.
This pull request also propagates ProgressMonitor to getNonAmbiguousGene and     getSampleByCancerStudyAndSampleId methods as an optional parameter, if it specified errors and warnings are stored in the ProgressMonitor, otherwise System.err is used.
ProgressMonitor now also writes the warnings to a log file as the output can be sometimes overwhelming the console.

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java [hunk 1]
```diff
@@ -338,7 +339,16 @@ public CanonicalGene getNonAmbiguousGene(String geneId) {
...
+    public CanonicalGene getNonAmbiguousGene(String geneId, String chr) { return getNonAmbiguousGene(geneId, chr, null); }
+
+        /**
+         * Look for gene that can be non-ambiguously determined.
+         * @param geneId an Entrez Gene ID or HUGO symbol or gene alias
+         * @param chr chromosome
+         * @param pMonitor for import warnings
+         * @return a gene that can be non-ambiguously determined, or null if cannot.
+         */
+    public CanonicalGene getNonAmbiguousGene(String geneId, String chr, ProgressMonitor pMonitor) {
```
Comment (jjgao) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java]: I don't like to put pMoniter in the function signature. Maybe we should have a ProgressMonitor singleton class that script and dao classes can call? And the singleton class is only initialized when importing (not the web code).

Comment (j-hudecek) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java]: I agree that singleton would be useful. The getUnambiguousGeneName is also called on tomcat so we have to make sure that case is also handled - only print the warning to the log


________________________________

Comment (j-hudecek) [inline core/src/main/java/org/mskcc/cbio/portal/dao/DaoGeneOptimized.java]: So I looked into making it a singleton and while it makes sense it is a really big change that touches almost a hundred of files and not all the changes are straightforward.
Are you sure you want me to do this?

For deciding whether we are running in a web server or not this method could be used: http://docs.liferay.com/portal/5.1/javadocs/portal-kernel/com/liferay/portal/kernel/util/ServerDetector.java.html

**Comment by j-hudecek** [comment]
So I've updated the pull request to make ProgressMonitor a singleton, what do you think of it now?

**Comment by jjgao** [comment]
Thanks @j-hudecek! I think it's great.

@zheins could you also review the change?

**Comment by zheins** [comment]
@j-hudecek I really like these changes, great job! @jjgao @n1zea144, this PR requires some changes in our pipelines, which I went ahead and made.

**Comment by zheins** [comment]
@j-hudecek, could you merge the conflicts?

**Comment by j-hudecek** [comment]
thanks, this should work right?

**Comment by fedde-s** [comment]
@pieterlukasse might want to see if this overlaps any of the work he planned or did in making the output of the import process' Java layer more user-friendly, as part of issue https://github.com/cBioPortal/cbioportal/issues/528

**Comment by fedde-s** [comment]
Looks great though!

**Comment by jjgao** [comment]
I think this looks great. Thanks, @j-hudecek! I'll merge it now.

@pieterlukasse if you have any further improvements, we can address them in another pull request.

**Comment by pieterlukasse** [comment]
@j-hudecek , @jjgao : awesome news, thanks! We actually have a client asking for improvements in this part, so I will definitely merge it in our branch and test it as well! I will make a new PR in case of improvements from our side.

**Pull request merged**